### ISSUE TITLE
feat: abstraction over filter and store for subscriptions with message recovery

### ISSUE THREAD

**Issue by adklempner** [waku-org/js-waku#2008]
This is a **bug report/feature request/support request/change request**

## Problem

If you run a filter subscription and go offline, any missed messages while offline will not be recovered. There can be other conditions that lead to missing messages from filter as well.

## Proposed Solutions

We should explore an abstraction over both filter and store that is able to:
- identify if any messages were missed
- use store to recover missed messages
- (?) have a single stream that supplies these messages from both sources
  - duplicate messages need to be detected and filtered out

## Notes

- Be aware of changes to store from v3

**Comment by weboko** [comment]
**Before attempting to implement** this feature - we need to know if pings(`store-v3` I assume) are light weight.

We need to be clear in the assumptions and check them:
- is `Store` reliable enough for us to be source of truth;
- (mentioned) how lightweight are pings;
- some other;

Side effect of the feature is: since right now Filter in `js-waku` and `status-go` has redundancy of 2-3 peers we might be able to drop this number to 1 with constantly checking if peer app subscribed to broadcasts all messages, if it is not the case - drop and replace with a new one.

**Comment by jm-clius** [comment]
Indeed. The Store v3 "ping" functionality is indeed there to have a (fairly) lightweight mechanism to retrieve only message hashes based on some query criteria.
We should consider "ephemeral" messages as best effort, also from a publishing POV - these messages will not make use of any Store-enhanced reliability mechanisms.
For now, assuming that Store is reliable is a reasonable abstraction for lightpush and filter.

My only comment would be that redundant Store checks should be avoided: lightpush can make use of regular Store checks to verify that a message has been published, filter can use the same checks to detect and retrieve missing received messages. In fact, this mechanism is agnostic as to what publish/subscribe methods are used and can be equally well applied to Relay.

**Comment by weboko** [comment]
Unblocked, moving to todo.
